test_runner_lib/include/test_runner/TestOutputSpec.h

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright (c) 2022 Trail of Bits, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <glog/logging.h>
+#include <remill/Arch/Arch.h>
+#include <test_runner/TestRunner.h>
+
+#include <any>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+namespace test_runner {
+
+using MemoryModifier = std::function<void(MemoryHandler &)>;
+
+struct RegisterPrecondition {
+  std::string register_name;
+  std::variant<uint64_t, uint8_t> enforced_value;
+};
+
+struct MemoryPostcondition {
+  uint64_t addr;
+  std::vector<uint8_t> bytes;
+};
+
+
+template <typename State, typename = typename std::enable_if_t<
+                              std::is_base_of<ArchState, State>::value>>
+class TestOutputSpec {
+ public:
+  uint64_t addr;
+  std::string target_bytes;
+
+ private:
+  remill::Instruction::Category expected_category;
+  std::vector<RegisterPrecondition> register_preconditions;
+  std::vector<RegisterPrecondition> register_postconditions;
+  std::vector<MemoryModifier> initial_memory_conditions;
+  std::vector<MemoryModifier> expected_memory_conditions;
+  std::unordered_map<std::string, std::function<std::any(State &)>>
+      reg_to_accessor;
+
+  template <typename T>
+  void ApplyCondition(State &state, std::string reg, T value) const {
+    auto accessor = reg_to_accessor.find(reg);
+    if (accessor != reg_to_accessor.end()) {
+      std::any_cast<std::reference_wrapper<T>>(accessor->second(state)).get() =
+          value;
+    }
+  }
+
+  template <typename T>
+  void CheckCondition(State &state, std::string reg, T value) const {
+    auto accessor = reg_to_accessor.find(reg);
+    if (accessor != reg_to_accessor.end()) {
+      auto actual =
+          std::any_cast<std::reference_wrapper<T>>(accessor->second(state));
+      LOG(INFO) << "Reg: " << reg << " Actual: " << std::hex
+                << static_cast<uint64_t>(actual.get())
+                << " Expected: " << std::hex << static_cast<uint64_t>(value);
```

Comment:
Maybe introduce `to_hex( ... )`? Seems like printing number of hex happens here an there, would hide some of the boiler casts.